### DISCUSSION TITLE
MPG PEDANT WITH 2 RASPBERRY PICOS

### DISCUSSION THREAD

**Discussion by M10CUBE** [grblHAL/RP2040#92] [Ideas]
I am working on this project since 2022 but only few hours spend on that. Here is the initial thoughts for the design
http://www.wizcube.eu/cnc.html
 The last few days I was looking what others did and start redesigning but stick on my initial idea to build using two Raspberry Pico.

General 
1 - Hand held desing
2 - Battery and charger for wireless portable use
3 - Two Raspberry Pico one with WiFi
4 - Design will be modular constructed in 2 PCBs following my motto "build once use it many times".  Both boards can be used for other projects individually.
5 - RS485 For wired connection powering the device as well
6 - WiFi for TCP/IP connection
7 - RX/TX communication
8 - Knob encoder input
9 - Keyboard input
10 - LCD
11 - KiCaD OSHW design
12 - Bluetooth communication (added) 

One PCB will be the encoder input  with Raspberry Pico without WiFi. You have the option to use three knob encoders or use one big commercial wheel and control the axes using keyboard switches. The rotary switch for axis and step selection used commercial devices is bulky and expensive. 

The other PCB will be used for communication, keypad and LCD
On next post I will publish the design so far and please comment any thoughts to make it better. It will be a dynamic design and I invite people to contribute knowing that many did excellent work on that in the past. I will find all the links I know and put it below.
Let us make the new generation of the Open source   MPG PEDANT. I am sure it will be useful to many of you.

![remote front](https://github.com/grblHAL/RP2040/assets/14018762/7fec7058-6dc3-4f0e-ad09-604d47890070)

[M10HM01-19.pdf](https://github.com/user-attachments/files/16067974/M10HM01-19.pdf)
[M10HM02-19.pdf](https://github.com/user-attachments/files/16067989/M10HM02-19.pdf)

**Comment by terjeio** [comment]
Where to start?

1. A network stack and a Telnet (or WebSocket?) client has to be added to the MPG code base. Some code from grblHAL can be reused. The grblHAL telnet daemon has to be extended to allow the MPG code to claim the telnet stream even if it is not connected, and to handle connects/disconnects cleanly.
It might be less work to go for Bluetooth instead assuming the controller is close to the pendant.

3. The keypad scanning code has to be ported to the second Pico.

4. A (breadboarded?) prototype has to be built to enable debugging and bring all the bits together?

1 or 2 is the most important to start with?

**Comment by M10CUBE** [comment]
Nicely put it @terjeio . Well let's start from the beginning.
I like your 123 approach.

1 - Bluetooth will be no problem on top of the other interfaces. (I did the edit on the first post). The Raspberry Pico has got it. One alternative module that can be used is the more advance ESP32-C6 found on a board from Waveshare and has got the same footprint as the Raspberry Pico. I strongly believe that Raspberry Pico team will come with something better on the new version. 
Having said that we stick on the Pico footprint since the future is bright (features, availability and economics speaking)
This part is mostly yours and can only help in software since my knowledge is mainly electronics and less in programming

2 - Yes . I can do that. I have some code from another project. 

3 - Breadboard is fine. I choose to partly build one using a wiring in a veroboard. A lot of work and I hate it. How about PCB milling the prototype. Do you thing double PCB milling is feasible?
Anyway I will finish the schematic and then do the PCB. 

Everything will be on line so make notes and give me directions for the hardware here:
https://drive.google.com/drive/folders/1Mo-kOKfVlPT8tJmywiodyOozxFDh6EGs?usp=sharing
M10HM01-19 The LCD and keyboard
M10HM02-19 The encoder input

Through the years I have collected some designs to help us here. If you have some more we can add it here.
The numbering is has only practical meaning.

1 - https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack
2 - https://openbuildspartstore.com/interface/
3 - https://github.com/SimonMerrett/grbl-MITM-pendant
4 - https://hackaday.io/project/27541-cnc-mpg

If you have some idea to invite others to help please do so.
